### ISSUE TITLE
Live reload breaks debugging session in elixirLS debugger

### ISSUE THREAD

**Issue by lukaszsamson** [phoenixframework/phoenix_live_reload#133]
I know that this most likely not be planned but there are repeated bugs and support requests related to that issue.

https://github.com/elixir-lsp/elixir-ls/issues/797
https://github.com/elixir-lsp/elixir-ls/issues/798

There are 2 problems when phoenix is run under elixirLS debugger
1. The code reloader (https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/code_reloader/server.ex) is purging and reloading modules on the fly with no regard for the ongoing debug session. The reloaded modules are not properly interpreted via erlang debugger.
2. It is messing with IO interception. Debug Adapter Protocol has strict requirements on stdout IO. All standard IO needs to be sent as JSONRPC protocol messages.

**Comment by josevalim** [comment]
This is tricky because I can’t come with good suggestions to address this. At best we could block the code reloader but I am not sure that would help either.

**Comment by josevalim** [comment]
@lukaszsamson if you can reproduce it we should do a pairing session :)

**Comment by coreyvan** [comment]
Is there any update to this issue or other workarounds? This makes the dev experience while debugging (I'm basically always debugging) pretty painful.

**Comment by lukaszsamson** [comment]
No. To address it either phoenix should check if modules are interpreted and call `:int.ni` after it recompiles or emit some event that the debug adapter can listen for and reinterpret. Probably breakpoints also need to be reset after that. A workaround could be to periodically reinterpret all modules in the VM that match launch config settings.

**Comment by CharlesIrvineKC** [comment]
My impression is that everyone in the Phoenix development community has given up on using a debugger and just use IO.inspect. Coming from a Java background where everyone used a debugger, I struggled for quite a while trying to get debugging working with Phoenix and ElixirLS and finally gave up and gave in to using IO.inspect. It's a shame I think. A lot of people would be very happy to be able to use the ElixirLS debugger with Phoenix.

**Comment by morganhein** [comment]
Ya this is a shame. I have been trying to audit phoenix and see if I should propose we switch to it for some things at the company I work for, but this is a hard blocker. I really love this ecosystem but being unable to use the tools like we're accustomed and requiring console logging is a non-starter.

That being said, thank you all that have worked on these projects and gotten things this far. Hopefully this gets fixed in the future, and then I'll revisit.

**Comment by CharlesIrvineKC** [comment]
I did a post for this on the forum. Let's see if anyone responds.

https://elixirforum.com/t/source-level-debugging-for-phoenix-and-phoenix-liveview-apps/65400
